examples/gateway/car-file-gateway/api.go

```diff
@@ -0,0 +1,234 @@
+		return nil, err
+	}
+
+	bsrv := blockservice.New(bs, offline.Exchange(bs))
+	dsrv := merkledag.NewDAGService(bsrv)
+
+	return &carGateway{
+		bs:   bs,
+		bsrv: bsrv,
+		dsrv: dsrv,
+	}, nil
+}
+
+func (api *carGateway) ResolvePath(ctx context.Context, p path.Path) (path.Resolved, error) {
+	if _, ok := p.(path.Resolved); ok {
+		return p.(path.Resolved), nil
+	}
+
+	if err := p.IsValid(); err != nil {
+		return nil, err
+	}
+
+	if p.Namespace() != "ipfs" {
+		return nil, fmt.Errorf("unsupported path namespace: %s", p.Namespace())
+	}
+
+	ipath := ipfspath.Path(p.String())
+
+	ipldFetcher := bsfetcher.NewFetcherConfig(api.bsrv)
+	ipldFetcher.PrototypeChooser = dagpb.AddSupportToChooser(func(lnk ipldPrime.Link, lnkCtx ipldPrime.LinkContext) (ipldPrime.NodePrototype, error) {
+		if tlnkNd, ok := lnkCtx.LinkNode.(schema.TypedLinkNode); ok {
+			return tlnkNd.LinkTargetNodePrototype(), nil
+		}
+		return basicnode.Prototype.Any, nil
+	})
+	fetcher := ipldFetcher.WithReifier(unixfsnode.Reify)
+	resolver := ipfspathresolver.NewBasicResolver(fetcher)
+
+	node, rest, err := resolver.ResolveToLastNode(ctx, ipath)
+	if err != nil {
+		return nil, err
+	}
+
+	root, err := cid.Parse(ipath.Segments()[1])
+	if err != nil {
+		return nil, err
+	}
+
+	return path.NewResolvedPath(ipath, node, root, gopath.Join(rest...)), nil
+}
+
+func (api *carGateway) resolveNode(ctx context.Context, p path.Path) (ipld.Node, error) {
+	rp, err := api.ResolvePath(ctx, p)
+	if err != nil {
+		return nil, err
+	}
+
+	node, err := api.dsrv.Get(ctx, rp.Cid())
+	if err != nil {
+		return nil, fmt.Errorf("get node: %w", err)
+	}
+	return node, nil
+}
+
+func (api *carGateway) GetUnixFsNode(ctx context.Context, p path.Resolved) (files.Node, error) {
+	nd, err := api.resolveNode(ctx, p)
+	if err != nil {
+		return nil, err
+	}
+
+	return unixfile.NewUnixfsFile(ctx, api.dsrv, nd)
+}
+
+func (api *carGateway) LsUnixFsDir(ctx context.Context, p path.Resolved) (<-chan iface.DirEntry, error) {
+	settings, err := options.UnixfsLsOptions(
+		options.Unixfs.ResolveChildren(false),
+		options.Unixfs.UseCumulativeSize(true),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	dagnode, err := api.resolveNode(ctx, p)
+	if err != nil {
+		return nil, err
+	}
+
+	dir, err := uio.NewDirectoryFromNode(api.dsrv, dagnode)
+	if err == uio.ErrNotADir {
+		return api.lsFromLinks(ctx, dagnode.Links(), settings)
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	return api.lsFromLinksAsync(ctx, dir, settings)
+}
+
+func (api *carGateway) processLink(ctx context.Context, linkres ft.LinkResult, settings *options.UnixfsLsSettings) iface.DirEntry {
+	if linkres.Err != nil {
+		return iface.DirEntry{Err: linkres.Err}
+	}
+
+	lnk := iface.DirEntry{
+		Name: linkres.Link.Name,
+		Cid:  linkres.Link.Cid,
+	}
+
+	switch lnk.Cid.Type() {
+	case cid.Raw:
+		// No need to check with raw leaves
+		lnk.Type = iface.TFile
+		lnk.Size = linkres.Link.Size
+	case cid.DagProtobuf:
+		if !settings.ResolveChildren {
+			break
+		}
```

Comment:
We know this is true which means basically all of the complexity around types can be removed.